kv: use generation in range descriptor cache and lookup

jeffrey-xiao opened the pull request:
This PR does the following things:

1. Add `generation_comparable` to the RangeDescriptor as an indication of when we can trust the generation. Prior to #36654, merges weren't considering the generation of the RHS and splits weren't setting the generation of the RHS. We want to ensure that the ranges are using the correct generation semantics.
2. Increment `generation` when changing replicas.
3. Update range lookup to use generations. When scanning across two meta ranges, it might return two committed, but overlapping ranges since it's not transactionally done. We can determine which range is stale by comparing the generation if both `generation_comparable` is set. If the generations are incomparable, we randomly choose a range descriptor. Previously we were always choosing the smaller range, but this is no longer accurate with range merges.
4. Update range descriptor cache insertion to use generations. Similar logic here as above. One difference with the new implementation is even if the range descriptor is not inserted (I.E. a overlapping range descriptor with a higher generation is already in the cache), it will still evict all overlapping range descriptors with lower generations.

Touches #10751.

I've collected metrics to compare this PR against master (`15d4329`) under two workloads while the cluster is running `kv50`.

1. Split workload: Two workers splitting random ranges: `ALTER TABLE kv.kv SPLIT AT VALUES (CAST(ROUND(RANDOM() * 9223372036854775808) AS INT)) WITH EXPIRATION '1s'`
2. Relocate workload: Two workers relocating random ranges: `ALTER TABLE kv.kv EXPERIMENTAL_RELOCATE SELECT ARRAY[%s, %s, %s], CAST(ROUND(RANDOM() * 9223372036854775808) AS INT)`. The new set of replicas is a random subset of the available replicas of size 3.

More details can be found in this [script](https://gist.github.com/jeffrey-xiao/2af9de98a56d8921dc7efa0cf7fb5b5e).

![image](https://user-images.githubusercontent.com/8853434/60835685-c7725700-a191-11e9-8303-c337dad424ba.png)

jeffrey-xiao committed:
storage: increment generation when changing replicas

The generation of a range descriptor can be used to determine whether a
range descriptor stored in the range descriptor cache is stale. Changing
the replica set of a range should increment the generation.

Release note: None

jeffrey-xiao committed:
kv: use generations in range descriptor cache and lookup

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38334)
<!-- Reviewable:end -->

nvb reviewed (commented):
This is really cool. Nice job. @tbg might be interested in taking a look as well.

> Reviewed 4 of 4 files at r1, 7 of 7 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-Li91mMA2vTQy2iO87Sq:b-levmz8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/internal/client/range_lookup.go, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li9-B_G-I5V1urKT1vb:-Li9-B_G-I5V1urKT1vc:bwd751x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c74d5b1cc0c41df7439f38a411303273bcf28125/pkg/internal/client/range_lookup.go#L233)):*
> ```Go
> 							// reservoir sampling. Note that we cannot determine the newer
> 							// version of the descriptor by looking at the size of the range
> 							// because both splits and merges can happen.
> ```

Explain that randomness is ok because if we guess wrong, we'll just try the lookup again. Randomness ensures that we'll probabilistically converge to the right descriptor.
___
*[pkg/kv/range_cache.go, line 636 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li90P1m3e_NoR7PGH8Y:-Li90P1m3e_NoR7PGH8Z:bcv10o9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c74d5b1cc0c41df7439f38a411303273bcf28125/pkg/kv/range_cache.go#L636)):*
> ```Go
> 			} else if desc.Equal(*descriptor) {
> 				// Generations are not comparable, but they are equal so don't evict.
> 				// TODO: Should this check for StartKey, EndKey, and replica set instead?
> ```

If they are `Equal` then won't those all be the same?

Also, if we're planning on leaving the TODO in then add an assignee. I think we have a lint for this.
___
*[pkg/kv/range_cache_test.go, line 1089 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li9154XAy2rxvKxb_8C:-Li9154XAy2rxvKxb_8D:b-obwm8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c74d5b1cc0c41df7439f38a411303273bcf28125/pkg/kv/range_cache_test.go#L1089)):*
> ```Go
> }
> 
> func TestRangeCacheGeneration(t *testing.T) {
> ```

Nice testing!
___
*[pkg/kv/range_cache_test.go, line 1097 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li91EnqDkXML3quSRoX:-Li91EnqDkXML3quSRoY:b-iej77k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c74d5b1cc0c41df7439f38a411303273bcf28125/pkg/kv/range_cache_test.go#L1097)):*
> ```Go
> 	}
> 
> 	initialDesc1 := &roachpb.RangeDescriptor{
> ```

This test would be easier to read if these descriptors were named something that described them. For instance, this could be named `descAM1`. The next could be named `descMZ3`. The final one could be `descAZIncomparable`.
___
*[pkg/roachpb/metadata.proto, line 194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8yS0E2agAx25uU_EF:-Li8yS0FD0BymI7DJzOU:b-f3ac3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b685884d4cc557714a476e71970757ab26aa4/pkg/roachpb/metadata.proto#L194)):*
> ```
> casuality
> ```

causality?
___
*[pkg/roachpb/metadata.proto, line 197 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8ymi-9bZbIz0HyYYt:-Li8ymi-9bZbIz0HyYYu:b-bqff1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b685884d4cc557714a476e71970757ab26aa4/pkg/roachpb/metadata.proto#L197)):*
> ```protobuf
>   // descriptors if they overlap and both their generation_comparable are set.
>   // If one or both of their generation_comparable is not set, we cannot
>   // determine which range descriptor is newer.
> ```

Make a note that this is needed because older versions of Cockroach didn't set the generation in some or all cases where it is currently set and so this allows code that relies on a total order of range descriptor generations to know when this order is ensured. It is a form of a migration.
___
*[pkg/storage/client_replica_test.go, line 1661 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8xWyGAMEcaggdeqlJ:-Li8xWyH9e7EslZIK2hT:b-owu37x) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/client_replica_test.go#L1661)):*
> ```Go
> 	}
> 
> 	oldGeneration := repl.Desc().GetGeneration()
> ```

Let's perform a `REMOVE_REPLICA` operation as well to make sure that that code path is tested as well.
___
*[pkg/storage/replica_command.go, line 925 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8xlIT1M41zvhFqC6L:-Li8xlIT1M41zvhFqC6M:bxf64m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/replica_command.go#L925)):*
> ```Go
> 		}
> 	}
> 	updatedDesc.IncrementGeneration()
> ```

nit: move this right below `updatedDesc := *desc` to mirror the structure in `adminSplitWithDescriptor` and `AdminMerge`.
___
*[pkg/storage/replica_command.go, line 248 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8zJcmCcG5hnnTdmSF:-Li8zJcn7SEKp_qMyG9A:b-aby2nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b685884d4cc557714a476e71970757ab26aa4/pkg/storage/replica_command.go#L248)):*
> ```Go
> 	leftDesc := *desc
> 	leftDesc.IncrementGeneration()
> 	leftDesc.GenerationComparable = true
> ```

We only want to set this is the cluster version is high enough such that it will never be lost. Let's pull out a `maybeMarkGenerationComparable` function in this file that takes a `*cluster.Settings` and checks whether a new cluster version is active before setting the flag. The function can also short-circuit before the settings check if the flag is already set.
___
*[pkg/storage/replica_command.go, line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li9-_dV4AWSPPmxHKmv:-Li9-_dV4AWSPPmxHKmw:boboxcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b685884d4cc557714a476e71970757ab26aa4/pkg/storage/replica_command.go#L255)):*
> ```Go
> 	// of why generations are useful.
> 	rightDesc.Generation = leftDesc.Generation
> 	rightDesc.GenerationComparable = true
> ```

We'll want to set this to true on the initial state of range descriptors during cluster bootstrapping too. See `store_bootstap.go`.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8y1j93XtWBDe4Vs7E:-Li8y1j93XtWBDe4Vs7F:b-m79dc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/batcheval/cmd_end_transaction.go#L1125)):*
> ```Go
> 	cpy.SetReplicas(roachpb.MakeReplicaDescriptors(change.UpdatedReplicas))
> 	cpy.NextReplicaID = change.NextReplicaID
> 	cpy.IncrementGeneration()
> ```

I'm surprised to see this here. Is it needed? The other two forms of descriptor modification only include the generation increment in their "admin" command and not in the resulting transaction trigger.


<!-- Sent from Reviewable.io -->

jeffrey-xiao committed:
roachpb: add generation_comparable to RangeDescriptor

Generations for range descriptors were introduced in a way where it was
impossible to determine which range descriptor was stale if they were
overlapping. This is additional state that we need to add to use the
generation. Two generations are comparable if both their
generation_comparable are set.

See #36654.

Release note: None

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8y1j93XtWBDe4Vs7E:-Li9wcGo7SKpIpzejPX6:b-mnv0j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/batcheval/cmd_end_transaction.go#L1125)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm surprised to see this here. Is it needed? The other two forms of descriptor modification only include the generation increment in their "admin" command and not in the resulting transaction trigger.
</blockquote></details>

Unlike the other admin commands, `ChangeReplicas` does not include the updated `RangeDescriptor` in the commit trigger, so you have to reconstruct it. See https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/replica_command.go#L971-L973. Will checking the cluster version here lead to a race?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Hey very nice! I was just talking about this again recently with @danhhz because we want to use generations to improve the replicaGC story when a snapshot arrives and finds an overlapping replica. After this PR, it seems that nothing's blocking us from doing so (at least if the involved ranges have the bit set).
The code looks good. This hasn't been touched for a quite a while, so any changes you can make to improve the commentary are appreciated (I pointed out a few, but there's always more).

> Reviewed 12 of 12 files at r4, 11 of 11 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LiDfDXp2XPt49vzw3oK:bpe75pt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/internal/client/range_lookup.go, line 227 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDSrPE5T49GtbTOo3s:-LiDSrPF6EPM8gBaGbj1:bi0z7v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/internal/client/range_lookup.go#L227)):*
> ```Go
> 						}
> 					} else {
> 						if rand.Intn(index+1) == 0 {
> ```

Why this change? Is there a known bug with the old approach?
Also the fact that this uses the global rand without making any test flaky (?) makes me thing this isn't tested properly.
___
*[pkg/kv/range_cache.go, line 610 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDcTmh4vF31GwaLIVf:-LiDcTmiDK4fzjNUXJRz:bmq4epp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L610)):*
> ```Go
> // in the cache; the bool return value specifies if the insertion should go on:
> // if the specified descriptor is already in the cache, then nothing is deleted
> // and false is returned. Otherwise, true is returned.
> ```

This isn't quite true. If any overlapping descriptor is known to be newer than the one passed in, false is returned, and true otherwise.
___
*[pkg/kv/range_cache.go, line 624 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDTyj965ChqK0tJaeW:-LiDTyj965ChqK0tJaeX:b-7yc0w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L624)):*
> ```Go
> 	entry, ok := rdc.rangeCache.cache.CeilEntry(rangeCacheKey(endMeta))
> 	if ok {
> 		descriptor := entry.Value.(*roachpb.RangeDescriptor)
> ```

While you're here consider s/descriptor/cached/.
___
*[pkg/kv/range_cache.go, line 625 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDX-hyC4yWtv00j8ka:-LiDX-hyC4yWtv00j8kb:b7owtln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L625)):*
> ```Go
> 	if ok {
> 		descriptor := entry.Value.(*roachpb.RangeDescriptor)
> 		if descriptor.StartKey.Less(desc.EndKey) && !descriptor.EndKey.Less(desc.EndKey) {
> ```

Am I reading it right that the condition here means "descriptor (cached) is a suffix of desc" (i.e. it may or may not touch the start key, but it definitely touches the end key)? Improve the commentary here (at CeilEntry above) a bit to make it more obvious that that's what it is and why that is the right thing, perhaps with a few inline examples.
___
*[pkg/kv/range_cache.go, line 641 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDbXiq-ifT6wnUzaAq:-LiDbXiq-ifT6wnUzaAr:bmpvkca) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L641)):*
Grammar here looks off. Suggestion

> // Clear any older descriptors overlapping desc. For example, if ranges [a,b) and [b,c) merged, old descriptors for both might be cached and we'd replace them when desc=[a,c).
// We can usually tell which descriptor is older based on the Generation. In the legacy case in which we can't, we clear unconditionally.
___
*[pkg/kv/range_cache_test.go, line 1137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdSAvBH5LUjroYYZP:-LiDdSAvBH5LUjroYYZQ:b-biupgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1137)):*
> ```Go
> 		expectedDesc []*roachpb.RangeDescriptor
> 	}{
> 		{
> ```

// descBY0 is ignored since the existing keyspace is covered by descriptors of generations 1 and 3, respectively.
___
*[pkg/kv/range_cache_test.go, line 1143 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdcX850yvwNsGp2ku:-LiDdcX850yvwNsGp2kv:beo5hnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1143)):*
> ```Go
> 			expectedDesc: []*roachpb.RangeDescriptor{descAM1, descMZ3},
> 		},
> 		{
> ```

// descBY2 evicts descAM1, but not descMZ3 based on Generation. Since it fails to evict the overlapping descriptor descMZ3 it doesn't insert itself.
___
*[pkg/kv/range_cache_test.go, line 1151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdvcr1ToXk1AEnwKB:-LiDdvcr1ToXk1AEnwKC:b-losrhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1151)):*
> ```Go
> 		{
> 			name:         "generation comparable evict 2",
> 			insertDesc:   descBY4,
> ```

// descBY4 replaces both existing descriptors and as a result ends up being the only descriptor left.
___
*[pkg/kv/range_cache_test.go, line 1157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDe2YACQJAw091Eep7:-LiDe2YACQJAw091Eep8:b93mif3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1157)):*
> ```Go
> 		{
> 			name:         "generation incomparable evict 2",
> 			insertDesc:   descBYIncomparable,
> ```

An incomparable descriptor evicts anything that gets in its way. This behavior is clearly less desirable in general but there's no better option in this case.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8y1j93XtWBDe4Vs7E:-LiCjD0A9EDdxaAgk6w6:b-3ey771) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/batcheval/cmd_end_transaction.go#L1125)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Unlike the other admin commands, `ChangeReplicas` does not include the updated `RangeDescriptor` in the commit trigger, so you have to reconstruct it. See https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/replica_command.go#L971-L973. Will checking the cluster version here lead to a race?
</blockquote></details>

You can use the cluster version here (this is above Raft) but that doesn't solve the problem that you have:

The caller may not have had the code to increment the generation, but the result will still have it bumped, but the intent (i.e. the other copies of the desc) won't have it; or the other way around. You might get away with this because likely (and you'd have to check) nothing uses `GenerationComparable` or `Generation` on this copy of the desc. But if you want to play it safe, you plumb the generation (and whether it's comparable) through instead.

On a related note, I'm curious whether (at the end of this PR) we have a mechanism that can upgrade a range descriptor to be comparable by basically carrying out a ChangeReplicas that doesn't actually change the set of members. Could you check what, if anything, prevents that today? We can then consider loosening the rules enough to allow this, so that by the next release we have at least the option to introduce some background process that upgrades all descriptors over time. Production clusters may have ranges that won't get split/merged/moved for years, and it'd be nice to migrate out of that at some point if we find that we want to get rid of the replicaGC queue completely.

On another related note, for replication changes we already have a sort of generation, namely `NextReplicaID`. It looks like theoretically we could stop using that in favor of `Generation`, but the migration is annoying and there's no real upshot, so just putting this out there and saying we shouldn't do it. But we could.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/internal/client/range_lookup.go, line 227 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDSrPE5T49GtbTOo3s:-LiFFvqI90qxHCHgPNhr:byjps02) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/internal/client/range_lookup.go#L227)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why this change? Is there a known bug with the old approach?
Also the fact that this uses the global rand without making any test flaky (?) makes me thing this isn't tested properly.
</blockquote></details>

The old approach assumed the range with the smaller key is the newer key, which isn't always the case if there are merges. I could add a test to cover the random selection, but I was wonder what that test should even test for?
___
*[pkg/kv/range_cache.go, line 610 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDcTmh4vF31GwaLIVf:-LiFm4HgCUlvUSChQDF5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L610)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This isn't quite true. If any overlapping descriptor is known to be newer than the one passed in, false is returned, and true otherwise.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache.go, line 624 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDTyj965ChqK0tJaeW:-LiFm4nO-gS3rEXiHEpF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L624)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

While you're here consider s/descriptor/cached/.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache.go, line 625 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDX-hyC4yWtv00j8ka:-LiFm52VAltwApbzpkPS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L625)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Am I reading it right that the condition here means "descriptor (cached) is a suffix of desc" (i.e. it may or may not touch the start key, but it definitely touches the end key)? Improve the commentary here (at CeilEntry above) a bit to make it more obvious that that's what it is and why that is the right thing, perhaps with a few inline examples.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache.go, line 641 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDbXiq-ifT6wnUzaAq:-LiFm5ywBFrUTHTfsn0Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache.go#L641)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Grammar here looks off. Suggestion

> // Clear any older descriptors overlapping desc. For example, if ranges [a,b) and [b,c) merged, old descriptors for both might be cached and we'd replace them when desc=[a,c).
// We can usually tell which descriptor is older based on the Generation. In the legacy case in which we can't, we clear unconditionally.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache_test.go, line 1137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdSAvBH5LUjroYYZP:-LiFm6c05Eh7osdi1-Wa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1137)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

// descBY0 is ignored since the existing keyspace is covered by descriptors of generations 1 and 3, respectively.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache_test.go, line 1143 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdcX850yvwNsGp2ku:-LiFm72p3hYNvXV1JAtL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1143)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

// descBY2 evicts descAM1, but not descMZ3 based on Generation. Since it fails to evict the overlapping descriptor descMZ3 it doesn't insert itself.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache_test.go, line 1151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDdvcr1ToXk1AEnwKB:-LiFm7UG7QfD2EFZUwZy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1151)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

// descBY4 replaces both existing descriptors and as a result ends up being the only descriptor left.
</blockquote></details>

Done.
___
*[pkg/kv/range_cache_test.go, line 1157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiDe2YACQJAw091Eep7:-LiFmTS_-1lr09IbWyYa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e591f35b97b8ba425d75da19e8f65f66ed9bc852/pkg/kv/range_cache_test.go#L1157)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

An incomparable descriptor evicts anything that gets in its way. This behavior is clearly less desirable in general but there's no better option in this case.
</blockquote></details>

Done.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Li8y1j93XtWBDe4Vs7E:-LiFmUzpC0HaeGMHRDnE:ba2jwyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/953daff43c13b777a346270dc3f9babb6b14f46c/pkg/storage/batcheval/cmd_end_transaction.go#L1125)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You can use the cluster version here (this is above Raft) but that doesn't solve the problem that you have:

The caller may not have had the code to increment the generation, but the result will still have it bumped, but the intent (i.e. the other copies of the desc) won't have it; or the other way around. You might get away with this because likely (and you'd have to check) nothing uses `GenerationComparable` or `Generation` on this copy of the desc. But if you want to play it safe, you plumb the generation (and whether it's comparable) through instead.

On a related note, I'm curious whether (at the end of this PR) we have a mechanism that can upgrade a range descriptor to be comparable by basically carrying out a ChangeReplicas that doesn't actually change the set of members. Could you check what, if anything, prevents that today? We can then consider loosening the rules enough to allow this, so that by the next release we have at least the option to introduce some background process that upgrades all descriptors over time. Production clusters may have ranges that won't get split/merged/moved for years, and it'd be nice to migrate out of that at some point if we find that we want to get rid of the replicaGC queue completely.

On another related note, for replication changes we already have a sort of generation, namely `NextReplicaID`. It looks like theoretically we could stop using that in favor of `Generation`, but the migration is annoying and there's no real upshot, so just putting this out there and saying we shouldn't do it. But we could.
</blockquote></details>

I decided that it was a good time to fix some of those TODOs :) I added the updated descriptor to `ChangeReplicasTrigger` so we don't have to plumb down that information.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 21 of 21 files at r7, 3 of 3 files at r8, 10 of 10 files at r9, 4 of 4 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao, @nvanbenschoten, and @tbg)

___
*[pkg/roachpb/data.proto, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiJu2ng8ISeiSZy3y_a:-LiJu2ng8ISeiSZy3y_b:be6ldi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0afa937bb777e0e60d9d9dd58118b46c7b9495b6/pkg/roachpb/data.proto#L168)):*
> ```protobuf
>   // The replica being modified.
>   ReplicaDescriptor replica = 2 [(gogoproto.nullable) = false];
>   // The updated range descriptor.
> ```

This is going to require a migration, right? We can be running in a cluster with nodes who will receive one of these that don't know about this new field and will be searching for the old ones. We can also be running in a cluster with nodes who will send us one of these with only the old ones populated.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Reviewed 21 of 21 files at r7, 3 of 3 files at r8, 10 of 10 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/kv/range_cache_test.go, line 1148 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiK7bve7G1c41S92wkP:-LiK7bve7G1c41S92wkQ:b-qemt85) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53915fe631e7914c7a3a1d010b4c61f27fce191/pkg/kv/range_cache_test.go#L1148)):*
> ```Go
> 			// descBY2 evicts descAM1, but not descMZ3 based on Generation. Since
> 			// there is an overlapping descriptor with higher Generation (descMZ3),
> 			// it is not inserted.;
> ```

nit: `;`
___
*[pkg/roachpb/data.proto, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiJu2ng8ISeiSZy3y_a:-LiK7t187yAuCew2ttdr:b-qg6592) ([raw file](https://github.com/cockroachdb/cockroach/blob/0afa937bb777e0e60d9d9dd58118b46c7b9495b6/pkg/roachpb/data.proto#L168)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is going to require a migration, right? We can be running in a cluster with nodes who will receive one of these that don't know about this new field and will be searching for the old ones. We can also be running in a cluster with nodes who will send us one of these with only the old ones populated.
</blockquote></details>

Yep. I guess the way to do this is to populate the old field as well as the new one. If the new one is set, use that, otherwise synthesize the desc from the old field (and generation will not be usable). Never populate the desc unless the cluster version has been bumped. The old fields are then retired in the next release.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
Hopefully I got all the aspects of the migration correct. PTAL.

- Both the change the `ChangeReplicasTrigger` and `RangeDescriptor` is version gated behind `VersionGenerationComparable`
- The updated descriptor will be reconstructed from `ChangeReplicasTrigger` if the updated descriptor is not specified, otherwise it's directly set.
- Incrementing `Generation` and setting `GenerationComparable` for `ChangeReplicas` is also version gated.
- `GenerationComparable` is a `* bool` that can be made non-nullable in 20.1

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LiO_mYJADBCMkpoHHgC:bzax1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Approach LGTM, just needs one more round to work out the last migration kinks.
@nvanbenschoten I saw some comment of you earlier about setting ourselves up for success to actually make these fields non-nullable in 20.1. To that end 19.2 (i.e. this code) will have to treat the `*zero` value and the nil value the same, right? Seems to be the case here.

> Reviewed 19 of 19 files at r11, 4 of 4 files at r12, 8 of 8 files at r13, 6 of 6 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LiPz-Ob3nuQzrpG81uF:b-hmy8rs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/internal/client/range_lookup_test.go, line 31 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPwutF1LhrtouVY4ig:-LiPwutG0Epo2ciipEV8:btm4cf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e331f8ffe2f0109049f7ae486c6d34975324b407/pkg/internal/client/range_lookup_test.go#L31)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	makeGeneration := func(generation int64) *int64 {
> ```

nit: `makeX` usually means that an `X` is returned, not a `*X`, so s/make/new/

Also can't you use `proto.Bool` and friends here like you do elsewhere?
___
*[pkg/storage/replica_command.go, line 1634 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPyJcL3YsUDr3yspgW:-LiPyJcM8z0pW4bjzwWI:b-q7nr1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e331f8ffe2f0109049f7ae486c6d34975324b407/pkg/storage/replica_command.go#L1634)):*
> ```Go
> // a high enough version such that GenerationComparable won't be lost.
> func (r *Replica) maybeMarkGenerationComparable(desc *roachpb.RangeDescriptor) {
> 	if !desc.GetGenerationComparable() && r.store.ClusterSettings().Version.IsActive(cluster.VersionGenerationComparable) {
> ```

The `!desc.GetGenerationComparable()` doesn't buy you anything.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPvqDq25K8YmaGfdeV:-LiPvqDr2A8DWT_r04OI:b-cdfwu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/95ada240f3bc1213d73b6dfe559e507f25ad7cab/pkg/storage/batcheval/cmd_end_transaction.go#L1118)):*
> ```Go
> 
> 	var desc roachpb.RangeDescriptor
> 	if rec.ClusterSettings().Version.IsActive(cluster.VersionGenerationComparable) {
> ```

`change.Desc` is only set if the cluster version is already active, so just make this `if change.Desc != nil`

This is simpler and also it avoids a problem: if the version check earlier populated `change.Desc` but now this node and this code here doesn't have the version active yet (could happen - propagates asynchronously) you throw away the original desc which contained strictly more info. As I argued elsewhere I don't *think* this causes any issues in this specific but who knows, let's play it safe.
Actually there's a second problem with the code as written: if the caller didn't have the setting (i.e. `change.Desc` is nil) but you have it here, you crash.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_command.go, line 1634 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPyJcL3YsUDr3yspgW:-LiQ0j548HG6JPuMCRem:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e331f8ffe2f0109049f7ae486c6d34975324b407/pkg/storage/replica_command.go#L1634)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The `!desc.GetGenerationComparable()` doesn't buy you anything.
</blockquote></details>

Done.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPvqDq25K8YmaGfdeV:-LiQ0w7D5dOIKR6-Jcg5:bgjzsg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/95ada240f3bc1213d73b6dfe559e507f25ad7cab/pkg/storage/batcheval/cmd_end_transaction.go#L1118)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`change.Desc` is only set if the cluster version is already active, so just make this `if change.Desc != nil`

This is simpler and also it avoids a problem: if the version check earlier populated `change.Desc` but now this node and this code here doesn't have the version active yet (could happen - propagates asynchronously) you throw away the original desc which contained strictly more info. As I argued elsewhere I don't *think* this causes any issues in this specific but who knows, let's play it safe.
Actually there's a second problem with the code as written: if the caller didn't have the setting (i.e. `change.Desc` is nil) but you have it here, you crash.
</blockquote></details>

Oops, I originally had a nil check for `change.Desc`, but I guess I messed up a rebase somewhere.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 14 of 14 files at r15, 4 of 4 files at r16, 8 of 8 files at r17, 4 of 4 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LiQ96wPCVLuPFHksNaN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/roachpb/metadata.proto, line 205 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiQ9_KWDXsLZKHRA6lC:-LiQ9_KWDXsLZKHRA6lD:br7v43j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88c80699e81800ab17a3291af4ae5b744341faa/pkg/roachpb/metadata.proto#L205)):*
> ```protobuf
>   // generations can determine when this order is ensured from this state. See
>   // #36654 for more details.
>   optional bool generation_comparable = 8;
> ```

Can you brush up `String()` so that it prints a `?` behind the generation in

https://github.com/cockroachdb/cockroach/blob/c097a16427f65e9070991f062716d222ea5903fe/pkg/roachpb/metadata.go#L243

if `!GetGenerationComparable()`?

That way we'll visually see if somehow we end up with non-comparable descriptors in clusters that should never see them.

I don't mind you doing this in a follow-up PR as this one has become difficult to review updates to.
___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 1118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiPvqDq25K8YmaGfdeV:-LiQ8n0tDOotp0IwlgWA:b8qhgws) ([raw file](https://github.com/cockroachdb/cockroach/blob/95ada240f3bc1213d73b6dfe559e507f25ad7cab/pkg/storage/batcheval/cmd_end_transaction.go#L1118)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Oops, I originally had a nil check for `change.Desc`, but I guess I messed up a rebase somewhere.
</blockquote></details>

Yeah, you also mentioned that in your post on the main thread. LG now


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 19 files at r11, 14 of 14 files at r15, 4 of 4 files at r16, 8 of 8 files at r17, 4 of 4 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/roachpb/data.proto, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiJu2ng8ISeiSZy3y_a:-Lii-56pCGvnhhTQ-yRf:b-eswell) ([raw file](https://github.com/cockroachdb/cockroach/blob/0afa937bb777e0e60d9d9dd58118b46c7b9495b6/pkg/roachpb/data.proto#L168)):*
> Never populate the desc unless the cluster version has been bumped.

Is this true? I would expect us to populate the desc regardless. If we didn't, a node sending one of these might not know that the cluster version has been bumped even after a 20.1 node joined the cluster. That would prevent us from removing the other fields in 20.1.
___
*[pkg/storage/replica_command.go, line 983 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-Lii-PMg5729d0K6jIX7:-Lii-PMg5729d0K6jIX8:b-7vd5az) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedf20d4484b58412030b6549622c21e812379f/pkg/storage/replica_command.go#L983)):*
> ```Go
> 			NextReplicaID:   updatedDesc.NextReplicaID,
> 		}
> 		if r.store.ClusterSettings().Version.IsActive(cluster.VersionGenerationComparable) {
> ```

See my comment above. I don't think we need to consider the cluster version at all for this.


<!-- Sent from Reviewable.io -->

jeffrey-xiao committed:
roachpb: add updated descriptor to ChangeReplicasTrigger

Previously, the receiver would have to reconstruct the updated range
descriptor using the updated_replicas and next_replica_id field in the
trigger. By using the updated descriptor instead of these fields, the
reconstruction can be avoided. Additionally, it was also easier to
change other fields in the descriptor without plumbing through the
commit trigger.

deprecated_updated_replicas and deprecated_next_replica_id must still be
included in ChangeReplicasTrigger for backwards compatibility, but they
can be removed in 20.1.

Release note: None

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @tbg)

___
*[pkg/roachpb/metadata.proto, line 205 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LiQ9_KWDXsLZKHRA6lC:-LioBbJw17bTNMDdKVEI:b-i3qh64) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88c80699e81800ab17a3291af4ae5b744341faa/pkg/roachpb/metadata.proto#L205)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can you brush up `String()` so that it prints a `?` behind the generation in

https://github.com/cockroachdb/cockroach/blob/c097a16427f65e9070991f062716d222ea5903fe/pkg/roachpb/metadata.go#L243

if `!GetGenerationComparable()`?

That way we'll visually see if somehow we end up with non-comparable descriptors in clusters that should never see them.

I don't mind you doing this in a follow-up PR as this one has become difficult to review updates to.
</blockquote></details>

Seemed like a pretty small change, so I just added it to this PR.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the two comments I left.

Nice job with the experiments here and in the other two PRs!

For others passing by: @jeffrey-xiao is planning on turning the experiment into a roachtest so that we can monitor range lookup efficiency and assert against regressions.

> Reviewed 16 of 16 files at r19, 4 of 4 files at r20, 9 of 9 files at r21, 4 of 4 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LjIJpJ6A89t7fGroumn:bwp8kk9)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jeffrey-xiao and @tbg)

___
*[pkg/roachpb/data.proto, line 171 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LjIJ8X36pmlIcgtoM_N:-LjIJ8X36pmlIcgtoM_O:bwqjwbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e608c2bc5755ca58fdfb620d25a8c3d2d3affe/pkg/roachpb/data.proto#L171)):*
> ```protobuf
>   repeated ReplicaDescriptor updated_replicas = 3 [(gogoproto.nullable) = false];
>   // The next replica id to use with this change applied.
>   int32 next_replica_id = 4 [(gogoproto.customname) = "NextReplicaID", (gogoproto.casttype) = "ReplicaID"];
> ```

Could you rename `updated_replicas` and `next_replica_id` to `deprecated_updated_replicas` and `deprecated_next_replica_id`?
___
*[pkg/storage/replica_command.go, line 977 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-LjIJRRO01MqgZTaxWQg:-LjIJRRO01MqgZTaxWQh:b-dvwo79) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e608c2bc5755ca58fdfb620d25a8c3d2d3affe/pkg/storage/replica_command.go#L977)):*
> ```Go
> 		}
> 
> 		trigger := &roachpb.ChangeReplicasTrigger{
> ```

Might as well inline this again.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38334#-:-LjMxnhuCDbvnPCsF_8q:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @tbg)



<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
TFTRs!

bors r+
